projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php

```diff
@@ -10,24 +10,95 @@ class Jetpack_Widget_Conditions {
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( isset( $_GET['legacy-widget-preview'] ) ) {
+			return;
+		}
+
+		// Not making use of the heartbeat functionality at all.
+		// phpcs:ignore WordPress.Security.NonceVerification.Missing
+		if ( isset( $_POST['action'] ) && 'heartbeat' === $_POST['action'] ) {
+			return;
+		}
+
+		// API call to *list* the widget types doesn't use editing visibility or display widgets.
+		if ( false !== strpos( $_SERVER['REQUEST_URI'], '/wp-json/wp/v2/widget-types?' ) ) {
+			return;
+		}
+
+		$add_data_assets_to_page = false;
+		$add_html_to_form        = false;
+		$handle_widget_updates   = false;
+
+		$using_classic_experience = ( ! function_exists( 'wp_use_widgets_block_editor' ) || ! wp_use_widgets_block_editor() );
+		if ( $using_classic_experience &&
+			(
+				is_customize_preview() || 'widgets.php' === $pagenow ||
+				// phpcs:ignore WordPress.Security.NonceVerification.Missing
+				( 'admin-ajax.php' === $pagenow && array_key_exists( 'action', $_POST ) && 'save-widget' === $_POST['action'] )
+			)
+		) {
+			$add_data_assets_to_page = true;
+			$add_html_to_form        = true;
+			$handle_widget_updates   = true;
+		} else {
+			// On a page that is hosting the API.
+			if ( is_customize_preview() || 'widgets.php' === $pagenow ) {
+				$add_data_assets_to_page = true;
+			}
+
+			// Encoding for a particular widget end point.
+			if ( 1 === preg_match( '|/wp-json/wp/v2/widget-types/.*/encode|', $_SERVER['REQUEST_URI'] ) ) {
+				$add_html_to_form      = true;
+				$handle_widget_updates = true;
+			}
+
+			// Batch API is usually saving but could be anything.
+			if ( false !== strpos( $_SERVER['REQUEST_URI'], '/wp-json/batch/v1' ) ) {
+				$handle_widget_updates = true;
+				$add_html_to_form      = true;
+			}
+
+			// Saving widgets via non-batch API. This isn't used within WordPress but could be used by third parties in theory.
+			if ( 'GET' !== $_SERVER['REQUEST_METHOD'] && false !== strpos( $_SERVER['REQUEST_URI'], '/wp/v2/widgets' ) ) {
+				$handle_widget_updates = true;
+				$add_html_to_form      = true;
+			}
+		}
+
+		if ( $add_html_to_form ) {
 			add_action( 'in_widget_form', array( __CLASS__, 'widget_conditions_admin' ), 10, 3 );
-		} elseif ( ! in_array( $GLOBALS['pagenow'], array( 'wp-login.php', 'wp-register.php' ) ) ) {
+		}
+
+		if ( $handle_widget_updates ) {
+			add_filter( 'widget_update_callback', array( __CLASS__, 'widget_update' ), 10, 3 );
+		}
+
+		if ( $add_data_assets_to_page ) {
+			add_action( 'sidebar_admin_setup', array( __CLASS__, 'widget_admin_setup' ) );
+		}
+
+		if ( ! $add_html_to_form && ! $handle_widget_updates && ! $add_data_assets_to_page &&
+			! in_array( $pagenow, array( 'wp-login.php', 'wp-register.php' ), true ) &&
+			// Don't filter widgets when editing them - otherwise they could get filtered out and become impossible to edit.
+			( false === strpos( wp_get_raw_referer(), '/wp-admin/widgets.php' ) )
```

Comment:
I may be missing something here, but wouldn't that mean that if I click on the site title in the admin bar to access my home page after editing widgets, the widgets would not be filtered then?

Reply:
This is a very good point :(

The problem I have is that I can't find a way to tell the difference between the API being used to fetch the widgets to display, and the API being used to fetch the list of widgets to edit - it's the same end point - [/wp-json/wp/v2/sidebars](https://make.wordpress.org/core/2021/06/29/rest-api-changes-in-wordpress-5-8/).

Unless there's something I've missed, the other option is to never filter the list at all.

Reply:
That's a good point. I don't have a great solution here either :\

@cpapazoglou Have you had the chance to look at this?

Reply:
At first I thought this was a problem with conditions like "Hide if User is logged in" but trying to reproduce it now and the widget still displays after saving them while being in `http://localhost/wp-admin/widgets.php`.

@dsas is there a chance that this is resolved and we don't need `( false === strpos( wp_get_raw_referer(), '/wp-admin/widgets.php' ) )`? Otherwise can you provide the exact steps to reproduce the `they could get filtered out and become impossible to edit.`?

Reply:
Yes, it definitely still happens, here are the exact steps, sorry for not providing them earlier:

1. Add legacy widget using /wp-admin/widgets.php
2. Set visibility to "Show if User is logged out"
3. Press update to save them
4. **Reload the page**
5. See that the newly added widget does not appear.

Screen recording here if that helps: https://d.pr/v/lJVdjv

Reply:
@jeherve We have solved this by changing the referer check to only be used if using the API.

This means if the API is used it will filter out widgets, unless it's called from the widgets admin page. This means they can still be edited there. 

The customizer is unaffected - it doesn't use the same sidebars API anyway.

The solution is perhaps a little bit of a hack, other alternatives we thought about were:
 * Never filter out widgets from the API i.e. eliminate the referer check. It didn't seem completely correct and might surprise people. Particularly if someone was using this API for something other than widget editing.
 * Get a change upstream so that when the API is called it explains why it's calling e.g. by adding a context=editing request var, we can then look for that request var and choose not to run the filters. This would take some time to coordinate.

Reply:
Something to note is security. I was worried about depending upon the referer to decide whether or not to filter out conditions, as that's totally in control of the client. The sidebar endpoint is only available to users with the edit_theme_options cap though, so it's ok 👍